compiler/rustc_mir_build/src/build/scope.rs

```diff
@@ -721,6 +721,42 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         self.cfg.terminate(block, source_info, TerminatorKind::Resume);
     }
 
+    /// Sets up the drops for explict tail calls.
+    ///
+    /// Unlike other kinds of early exits, tail calls do not go through the drop tree.
+    /// Instead, all scheduled drops are immediately added to the CFG.
+    pub(crate) fn break_for_tail_call(&mut self, mut block: BasicBlock) -> BlockAnd<()> {
+        // the innermost scope contains only the destructors for the tail call arguments
+        // we only want to drop these in case of a panic, so we skip it
+        for scope in self.scopes.scopes[1..].iter().rev().skip(1) {
+            for drop in scope.drops.iter().rev() {
+                match drop.kind {
+                    DropKind::Value => {
+                        let target = self.cfg.start_new_block();
+                        let terminator = TerminatorKind::Drop {
+                            target,
+                            // The caller will handle this if needed.
```

Comment:
I don't think this is true here (looking at the MIR opt tests)

Reply:
I think actually the inner loop should be replaced with `build_scope_drops`. That should handle the unwinds correctly and reduce code duplication.

Reply:
I've picked @DrMeepster's fix commit (https://github.com/DrMeepster/rust/commit/f79affd8c3b08c70f5a248ecb48280e0310aafab) as af3f2f84d34cd8397e2ce36d91541acd38b1e223.

@matthewjasper does the code look correct now?